File: src-tauri/crates/yss-data-contract/src/value_type.rs
use serde::{Deserialize, Serialize};
use std::fmt;
use std::str::FromStr;
use thiserror::Error;

/// Structure of an analysis value. Scalar meaning belongs exclusively to SemanticType;
/// physical representation is carried by the value or field metadata, never by this type.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "kind", content = "inner")]
pub enum ValueType {
    Scalar(crate::SemanticType),
    Array(Box<ValueType>),
    Object,
    DataFrame,
    DataSeries(Box<ValueType>),
    Struct(std::string::String),
    OneOf(Vec<ValueType>),
    Any,
}

impl ValueType {
    /// Constructs a flattened union while preserving the first occurrence order.
    pub fn one_of(types: Vec<ValueType>) -> ValueType {
        let mut flat = Vec::new();
        for data_type in types {
            match data_type {
                ValueType::Any => return ValueType::Any,
                ValueType::OneOf(inner) => {
                    for item in inner {
                        if item == ValueType::Any {
                            return ValueType::Any;
                        }
                        if !flat.contains(&item) {
                            flat.push(item);
                        }
                    }
                }
                other => {
                    if !flat.contains(&other) {
                        flat.push(other);
                    }
                }
            }
        }

        match flat.len() {
            0 => ValueType::Any,
            1 => flat.pop().unwrap_or(ValueType::Any),
            _ => ValueType::OneOf(flat),
        }
    }

    /// Constructs the Numeric scalar meaning without selecting a physical representation.
    pub fn number() -> ValueType {
        ValueType::Scalar(crate::SemanticType::Numeric)
    }
}

impl fmt::Display for ValueType {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ValueType::Scalar(semantic) => write!(formatter, "{semantic}"),
            ValueType::Array(inner) => write!(formatter, "Array<{inner}>"),
            ValueType::Object => write!(formatter, "Object"),
            ValueType::DataFrame => write!(formatter, "DataFrame"),
            ValueType::DataSeries(inner) => write!(formatter, "DataSeries<{inner}>"),
            ValueType::Struct(key) => write!(formatter, "Struct<{key}>"),
            ValueType::OneOf(types) => {
                for (index, data_type) in types.iter().enumerate() {
                    if index > 0 {
                        write!(formatter, " | ")?;
                    }
                    write!(formatter, "{data_type}")?;
                }
                Ok(())
            }
            ValueType::Any => write!(formatter, "Any"),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Error)]
pub enum ValueTypeParseError {
    #[error("data type is empty")]
    Empty,
    #[error("unknown data type")]
    UnknownKind,
    #[error("malformed composite data type")]
    MalformedComposite,
}

impl FromStr for ValueType {
    type Err = ValueTypeParseError;

    fn from_str(source: &str) -> Result<Self, Self::Err> {
        let trimmed = source.trim();
        if trimmed.is_empty() {
            return Err(ValueTypeParseError::Empty);
        }

        let parts = split_top_level(trimmed, '|')?;
        if parts.len() > 1 {
            let types = parts
                .into_iter()
                .map(str::parse)
                .collect::<Result<Vec<_>, _>>()?;
            return Ok(ValueType::one_of(types));
        }

        if let Ok(semantic) = trimmed.parse::<crate::SemanticType>() {
            return Ok(ValueType::Scalar(semantic));
        }
        match trimmed {
            "Object" => Ok(ValueType::Object),
            "DataFrame" => Ok(ValueType::DataFrame),
            "DataSeries" => Ok(ValueType::DataSeries(Box::new(ValueType::Any))),
            "Any" => Ok(ValueType::Any),
            _ => parse_composite(trimmed),
        }
    }
}

fn parse_composite(source: &str) -> Result<ValueType, ValueTypeParseError> {
    if let Some(inner) = delimited_inner(source, "Array")? {
        let data_type = inner
            .parse()
            .map_err(|_| ValueTypeParseError::MalformedComposite)?;
        return Ok(ValueType::Array(Box::new(data_type)));
    }
    if let Some(inner) = delimited_inner(source, "DataSeries")? {
        let data_type = inner
            .parse()
            .map_err(|_| ValueTypeParseError::MalformedComposite)?;
        return Ok(ValueType::DataSeries(Box::new(data_type)));
    }
    if let Some(key) = delimited_inner(source, "Struct")? {
        return Ok(ValueType::Struct(key.to_owned()));
    }
    if source.contains(['<', '>']) {
        return Err(ValueTypeParseError::MalformedComposite);
    }
    Err(ValueTypeParseError::UnknownKind)
}

fn delimited_inner<'a>(
    source: &'a str,
    kind: &str,
) -> Result<Option<&'a str>, ValueTypeParseError> {
    let Some(rest) = source.strip_prefix(kind) else {
        return Ok(None);
    };
    if !rest.starts_with('<') || !rest.ends_with('>') {
        return Err(ValueTypeParseError::MalformedComposite);
    }
    let inner = &rest[1..rest.len() - 1];
    if !angles_are_balanced(inner) {
        return Err(ValueTypeParseError::MalformedComposite);
    }
    Ok(Some(inner))
}

fn split_top_level(source: &str, separator: char) -> Result<Vec<&str>, ValueTypeParseError> {
    let mut parts = Vec::new();
    let mut depth = 0usize;
    let mut start = 0;
    for (index, character) in source.char_indices() {
        match character {
            '<' => depth += 1,
            '>' => {
                depth = depth
                    .checked_sub(1)
                    .ok_or(ValueTypeParseError::MalformedComposite)?;
            }
            character if character == separator && depth == 0 => {
                let part = source[start..index].trim();
                if part.is_empty() {
                    return Err(ValueTypeParseError::MalformedComposite);
                }
                parts.push(part);
                start = index + character.len_utf8();
            }
            _ => {}
        }
    }
    if depth != 0 {
        return Err(ValueTypeParseError::MalformedComposite);
    }

    let tail = source[start..].trim();
    if tail.is_empty() {
        return Err(ValueTypeParseError::MalformedComposite);
    }
    parts.push(tail);
    Ok(parts)
}

fn angles_are_balanced(source: &str) -> bool {
    let mut depth = 0usize;
    for character in source.chars() {
        match character {
            '<' => depth += 1,
            '>' => {
                let Some(next) = depth.checked_sub(1) else {
                    return false;
                };
                depth = next;
            }
            _ => {}
        }
    }
    depth == 0
}

File: src/shared/types/domain/valueType.ts
import type { DeepReadonly } from "@/shared/types/deepReadonly";
import { SEMANTIC_TYPES, type SemanticType } from "./database";

/** Analysis meaning and structure; physical storage stays in field/value metadata. */
export type ValueType =
  | { kind: "Scalar"; inner: SemanticType }
  | { kind: "Object" }
  | { kind: "Any" }
  | { kind: "DataFrame" }
  | { kind: "Array"; inner: ValueType }
  | { kind: "DataSeries"; inner: ValueType }
  | { kind: "Struct"; inner: string }
  | { kind: "OneOf"; inner: ValueType[] };

export function isBackendDataType(value: unknown): value is ValueType {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const record = value as Record<string, unknown>;
  if (["Object", "Any", "DataFrame"].includes(record.kind as string))
    return Object.keys(record).length === 1;
  if (Object.keys(record).length !== 2 || !("inner" in record)) return false;
  if (record.kind === "Scalar") return SEMANTIC_TYPES.includes(record.inner as SemanticType);
  if (record.kind === "Struct")
    return typeof record.inner === "string" && record.inner.trim() !== "";
  if (record.kind === "Array" || record.kind === "DataSeries")
    return isBackendDataType(record.inner);
  return (
    record.kind === "OneOf" &&
    Array.isArray(record.inner) &&
    record.inner.length > 0 &&
    record.inner.every(isBackendDataType)
  );
}

export const CONSTANT_SELECTABLE_DATA_TYPE_KINDS = [
  ...SEMANTIC_TYPES,
  "DataFrame",
  "DataSeries",
] as const;
export const DATA_SERIES_ELEMENT_TYPE_KINDS = SEMANTIC_TYPES;
export const DEFAULT_ARRAY_VALUE: readonly number[] = [];
export const DEFAULT_OBJECT_VALUE: Readonly<Record<string, number>> = {};

export function dataTypeKind(type: ValueType): string {
  return type.kind === "Scalar" ? type.inner : type.kind;
}

export function dataTypeDisplay(type: DeepReadonly<ValueType>): string {
  switch (type.kind) {
    case "Scalar":
      return type.inner;
    case "Array":
    case "DataSeries":
      return `${type.kind}<${dataTypeDisplay(type.inner)}>`;
    case "Struct":
      return `Struct<${type.inner}>`;
    case "OneOf":
      return type.inner.map(dataTypeDisplay).join(" | ");
    default:
      return type.kind;
  }
}

function splitTopLevel(source: string): string[] | null {
  let depth = 0;
  let start = 0;
  const parts: string[] = [];
  for (let i = 0; i < source.length; i++) {
    if (source[i] === "<") depth++;
    else if (source[i] === ">" && --depth < 0) return null;
    else if (source[i] === "|" && depth === 0) {
      if (!source.slice(start, i).trim()) return null;
      parts.push(source.slice(start, i).trim());
      start = i + 1;
    }
  }
  if (depth || !source.slice(start).trim()) return null;
  parts.push(source.slice(start).trim());
  return parts;
}

function oneOf(types: ValueType[]): ValueType {
  const flat: ValueType[] = [];
  for (const type of types) {
    if (type.kind === "Any") return type;
    for (const member of type.kind === "OneOf" ? type.inner : [type]) {
      if (member.kind === "Any") return member;
      if (!flat.some((item) => dataTypeDisplay(item) === dataTypeDisplay(member)))
        flat.push(member);
    }
  }
  return flat.length === 0
    ? { kind: "Any" }
    : flat.length === 1
      ? flat[0]!
      : { kind: "OneOf", inner: flat };
}

export function dataTypeFromDisplayString(source: string): ValueType | null {
  const trimmed = source.trim();
  const parts = splitTopLevel(trimmed);
  if (!parts) return null;
  if (parts.length > 1) {
    const types = parts.map(dataTypeFromDisplayString);
    return types.every((type): type is ValueType => type !== null) ? oneOf(types) : null;
  }
  if (SEMANTIC_TYPES.includes(trimmed as SemanticType))
    return { kind: "Scalar", inner: trimmed as SemanticType };
  if (trimmed === "Any" || trimmed === "Object" || trimmed === "DataFrame")
    return { kind: trimmed };
  if (trimmed === "DataSeries") return { kind: "DataSeries", inner: { kind: "Any" } };
  const series = trimmed.match(/^(Array|DataSeries)<(.+)>$/);
  if (series) {
    const inner = dataTypeFromDisplayString(series[2]!);
    return inner ? { kind: series[1] as "Array" | "DataSeries", inner } : null;
  }
  const named = trimmed.match(/^Struct<([^<>]+)>$/);
  return named ? { kind: "Struct", inner: named[1]! } : null;
}

export function dataTypeFromKey(key: string, inner?: ValueType | string): ValueType {
  if (SEMANTIC_TYPES.includes(key as SemanticType))
    return { kind: "Scalar", inner: key as SemanticType };
  if (key === "Array" || key === "DataSeries")
    return {
      kind: key,
      inner: typeof inner === "object" ? inner : { kind: "Scalar", inner: "Numeric" },
    };
  if (key === "Struct") return { kind: "Struct", inner: typeof inner === "string" ? inner : "" };
  return dataTypeFromDisplayString(key) ?? { kind: "Any" };
}

export function isPrimitiveType(type: ValueType): boolean {
  return (
    type.kind === "Scalar" ||
    (type.kind === "OneOf" && type.inner.length > 0 && type.inner.every(isPrimitiveType))
  );
}

export function isComplexType(type: ValueType): boolean {
  return (
    ["DataFrame", "DataSeries", "Object", "Array"].includes(type.kind) ||
    (type.kind === "OneOf" && type.inner.length > 0 && type.inner.every(isComplexType))
  );
}

export function getDefaultValue(type: ValueType): unknown {
  if (type.kind === "Scalar")
    return type.inner === "Numeric" ? 0 : type.inner === "Binary" ? false : "";
  if (type.kind === "Array") return [];
  if (type.kind === "Object") return {};
  return undefined;
}
